src/main/java/frc/robot/subsystems/SwerveDriveSubsystem.java

```diff
@@ -0,0 +1,140 @@
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.common.Odometry;
+
+public class SwerveDriveSubsystem extends SubsystemBase{
+
+    private WheelSubsystem backRight;
+    private WheelSubsystem backLeft;
+    private WheelSubsystem frontRight;
+    private WheelSubsystem frontLeft;
+    
+    private final NetworkTableInstance ntInstance = NetworkTableInstance.getDefault();
+    private final NetworkTable table = ntInstance.getTable("/components/drivetrain");
+
+    private final NetworkTableEntry backRightAngleEncoder = table.getEntry("backRightAngleEncoder");
+    private final NetworkTableEntry backLeftAngleEncoder = table.getEntry("backLeftAngleEncoder");
+    private final NetworkTableEntry frontRightAngleEncoder = table.getEntry("frontRightAngleEncoder");
+    private final NetworkTableEntry frontLeftAngleEncoder = table.getEntry("frontLeftAngleEncoder");
+
+    private final NetworkTableEntry isFieldCentric = table.getEntry("isFieldCentric");
+
+    private final NetworkTable odometryTable = ntInstance.getTable("/common/Odometry");
+    private final NetworkTableEntry odometryPose = odometryTable.getEntry("odometryPose");
+    private final NetworkTableEntry velocity = table.getEntry("wheelvelocity");
+
+
+
+    private SwerveDriveKinematics kinematics;
+    private Odometry odometry;
+    
+    public boolean fieldCentric = false;
+
+    public SwerveDriveSubsystem (WheelSubsystem backRight, WheelSubsystem backLeft, WheelSubsystem frontRight, WheelSubsystem frontLeft, SwerveDriveKinematics kinematics, Odometry odometry) {
+        this.backRight = backRight;
+        this.backLeft = backLeft;
+        this.frontRight = frontRight;
+        this.frontLeft = frontLeft;
+
+        this.kinematics = kinematics;
+        this.odometry = odometry;
+
+        this.isFieldCentric.setBoolean(fieldCentric);
+    }
+
+
+    public void drive (double x, double y, double rot) {
+        
+
+        ChassisSpeeds speeds = new ChassisSpeeds(x, y, rot);
+        // ChassisSpeeds speeds = new ChassisSpeeds(0, 0, 0);
+
+        if (fieldCentric) {
+            speeds = ChassisSpeeds.fromFieldRelativeSpeeds(speeds, odometry.getRotation2d());
+        }
+        // Convert to module states
+        SwerveModuleState[] moduleStates = kinematics.toSwerveModuleStates(speeds);
+
+        SwerveModuleState frontLeftState = moduleStates[0];
+        SwerveModuleState frontRightState = moduleStates[1];
+        SwerveModuleState backLeftState = moduleStates[2];
+        SwerveModuleState backRightState = moduleStates[3];
+
+        velocity.setDouble(frontLeft.getSpeedMotor().getEncoder().getVelocity());
+
+        frontLeft.drive(frontLeftState);
+        frontRight.drive(frontRightState);
+        backLeft.drive(backLeftState);
+        backRight.drive(backRightState);
```

Comment:
you should use the drive method you have so that any time you want to make logic changes from moduleStates on you only have to once